### ISSUE TITLE
AMP much slower with groupped Conv2d

### ISSUE THREAD

**Issue by buoyancy99** [pytorch/pytorch#43544]
It seems automatic mix precision on conv2d with groups>1 is significantly slower than fp32
This result is more obvious on machines with Volta GPUs with Tensor cores.

```
import torch
import torch.nn as nn
from torch.cuda.amp import autocast
import time


class amp_net(nn.Module):
    def __init__(self, amp, group):
        super().__init__()
        self.amp = amp
        if group:
            self.net = nn.Sequential(nn.Conv2d(12, 64, kernel_size=3, stride=2, padding=1, groups=4),
                                     nn.ReLU(),
                                     nn.Conv2d(64, 128, kernel_size=3, stride=2, padding=1, groups=4),
                                     nn.ReLU(),
                                     nn.Conv2d(128, 256, kernel_size=3, stride=2, padding=1, groups=4),
                                     nn.ReLU(),
                                     nn.Conv2d(256, 512, kernel_size=3, stride=2, padding=1, groups=4),
                                     nn.ReLU(),
                                     nn.Conv2d(512, 1024, kernel_size=3, stride=2, padding=1, groups=4),
                                     nn.Flatten(),
                                     nn.Linear(9216, 5)
                                     ).cuda()
        else:
            self.net = nn.Sequential(nn.Conv2d(12, 64, kernel_size=3, stride=2, padding=1),
                                     nn.ReLU(),
                                     nn.Conv2d(64, 128, kernel_size=3, stride=2, padding=1),
                                     nn.ReLU(),
                                     nn.Conv2d(128, 256, kernel_size=3, stride=2, padding=1),
                                     nn.ReLU(),
                                     nn.Conv2d(256, 512, kernel_size=3, stride=2, padding=1),
                                     nn.ReLU(),
                                     nn.Conv2d(512, 1024, kernel_size=3, stride=2, padding=1),
                                     nn.Flatten(),
                                     nn.Linear(9216, 5)
                                     ).cuda()

    def forward(self, x):
        with autocast(self.amp):
            return self.net(x)


def test_with_arch(group):
    net_amp = amp_net(True, group)
    optimizer_amp = torch.optim.Adam(net_amp.parameters(), lr=1e-3)
    net = amp_net(False, group)
    optimizer = torch.optim.Adam(net.parameters(), lr=1e-3)
    amp_scaler = torch.cuda.amp.GradScaler()

    sample_input = torch.rand((64, 12, 84, 84)).cuda()
    sample_output = torch.randint(5, (64, )).cuda()
    criteria = nn.CrossEntropyLoss()

    start_time = time.time()
    for _ in range(50):
        with torch.cuda.amp.autocast(True):
            result = net_amp(sample_input)
            loss = criteria(result, sample_output)
        optimizer_amp.zero_grad()
        amp_scaler.scale(loss).backward()
        amp_scaler.unscale_(optimizer_amp)
        amp_scaler.step(optimizer_amp)
        amp_scaler.update()

    end_time = time.time()
    time_amp = end_time - start_time

    start_time = time.time()
    for _ in range(50):
        with torch.cuda.amp.autocast(False):
            result = net(sample_input)
            loss = criteria(result, sample_output)
        optimizer.zero_grad()
        amp_scaler.scale(loss).backward()
        amp_scaler.unscale_(optimizer)
        amp_scaler.step(optimizer)
        amp_scaler.update()
    end_time = time.time()
    time_fp32 = end_time - start_time

    prefix = 'Group Conv2d: ' if group else 'Conv2d: '
    print(prefix + 'amp time / fp32 time {}'.format(time_amp / time_fp32))


if __name__ == '__main__':
    test_with_arch(False)
    test_with_arch(True)

```

Result:
```
Conv2d: amp time / fp32 time 0.3734461807645061
Group Conv2d: amp time / fp32 time 3.4929036465152117
```

**Comment by mruberry** [comment]
Thank you for reporting this issue, @buoyancy99. @mcarilli to comment.

**Comment by ngimel** [comment]
You are not properly timing your code. Please use torch benchmarking utilities https://github.com/pytorch/pytorch/tree/master/torch/utils/_benchmark. Even without that, by slightly modifying your code to exclude warm-up and properly synchronize, I'm getting:
```
    for i in range(50):
        if i==4:
            torch.cuda.synchronize()
            start_time = time.time()
        with torch.cuda.amp.autocast(True):
            result = net_amp(sample_input)
            loss = criteria(result, sample_output)
        optimizer_amp.zero_grad()
        amp_scaler.scale(loss).backward()
        amp_scaler.unscale_(optimizer_amp)
        amp_scaler.step(optimizer_amp)
        amp_scaler.update()

    torch.cuda.synchronize()
    end_time = time.time()
    time_amp = end_time - start_time
```
```
Conv2d: amp time / fp32 time 0.27305754369762736
Group Conv2d: amp time / fp32 time 0.7220136417913702
```
Speed-up for grouped convolutions is expected to be much smaller than for regular.

**Comment by buoyancy99** [comment]
> You are not properly timing your code. Please use torch benchmarking utilities https://github.com/pytorch/pytorch/tree/master/torch/utils/_benchmark. Even without that, by slightly modifying your code to exclude warm-up and properly synchronize, I'm getting:
> 
> ```
>     for i in range(50):
>         if i==4:
>             torch.cuda.synchronize()
>             start_time = time.time()
>         with torch.cuda.amp.autocast(True):
>             result = net_amp(sample_input)
>             loss = criteria(result, sample_output)
>         optimizer_amp.zero_grad()
>         amp_scaler.scale(loss).backward()
>         amp_scaler.unscale_(optimizer_amp)
>         amp_scaler.step(optimizer_amp)
>         amp_scaler.update()
> 
>     torch.cuda.synchronize()
>     end_time = time.time()
>     time_amp = end_time - start_time
> ```
> 
> ```
> Conv2d: amp time / fp32 time 0.27305754369762736
> Group Conv2d: amp time / fp32 time 0.7220136417913702
> ```
> 
> Speed-up for grouped convolutions is expected to be much smaller than for regular.

Is this result on RTX gpu? I tried your code and got result:
```
Conv2d:amp time: 0.414s, fp32 time: 1.070s, amp/fp32: 0.387
Group Conv2d: amp time: 1.706s, fp32 time: 0.503s, amp/fp32: 3.394
```

Updated code:
```
import torch
import torch.nn as nn
from torch.cuda.amp import autocast
import time


class amp_net(nn.Module):
    def __init__(self, amp, group):
        super().__init__()
        self.amp = amp
        if group:
            self.net = nn.Sequential(nn.Conv2d(12, 64, kernel_size=3, stride=2, padding=1, groups=4),
                                     nn.ReLU(),
                                     nn.Conv2d(64, 128, kernel_size=3, stride=2, padding=1, groups=4),
                                     nn.ReLU(),
                                     nn.Conv2d(128, 256, kernel_size=3, stride=2, padding=1, groups=4),
                                     nn.ReLU(),
                                     nn.Conv2d(256, 512, kernel_size=3, stride=2, padding=1, groups=4),
                                     nn.ReLU(),
                                     nn.Conv2d(512, 1024, kernel_size=3, stride=2, padding=1, groups=4),
                                     nn.Flatten(),
                                     nn.Linear(9216, 5)
                                     ).cuda()
        else:
            self.net = nn.Sequential(nn.Conv2d(12, 64, kernel_size=3, stride=2, padding=1),
                                     nn.ReLU(),
                                     nn.Conv2d(64, 128, kernel_size=3, stride=2, padding=1),
                                     nn.ReLU(),
                                     nn.Conv2d(128, 256, kernel_size=3, stride=2, padding=1),
                                     nn.ReLU(),
                                     nn.Conv2d(256, 512, kernel_size=3, stride=2, padding=1),
                                     nn.ReLU(),
                                     nn.Conv2d(512, 1024, kernel_size=3, stride=2, padding=1),
                                     nn.Flatten(),
                                     nn.Linear(9216, 5)
                                     ).cuda()

    def forward(self, x):
        with autocast(self.amp):
            return self.net(x)


def test_with_arch(group):
    net_amp = amp_net(True, group)
    optimizer_amp = torch.optim.Adam(net_amp.parameters(), lr=1e-3)
    net = amp_net(False, group)
    optimizer = torch.optim.Adam(net.parameters(), lr=1e-3)
    amp_scaler = torch.cuda.amp.GradScaler()

    sample_input = torch.rand((64, 12, 84, 84)).cuda()
    sample_output = torch.randint(5, (64, )).cuda()
    criteria = nn.CrossEntropyLoss()

    for i in range(50):
        if i == 4:
            torch.cuda.synchronize()
            start_time = time.time()

        with torch.cuda.amp.autocast(True):
            result = net_amp(sample_input)
            loss = criteria(result, sample_output)
        optimizer_amp.zero_grad()
        amp_scaler.scale(loss).backward()
        amp_scaler.unscale_(optimizer_amp)
        amp_scaler.step(optimizer_amp)
        amp_scaler.update()

    torch.cuda.synchronize()
    end_time = time.time()
    time_amp = end_time - start_time

    start_time = time.time()
    for i in range(50):
        if i == 4:
            torch.cuda.synchronize()
            start_time = time.time()
        with torch.cuda.amp.autocast(False):
            result = net(sample_input)
            loss = criteria(result, sample_output)
        optimizer.zero_grad()
        amp_scaler.scale(loss).backward()
        amp_scaler.unscale_(optimizer)
        amp_scaler.step(optimizer)
        amp_scaler.update()

    torch.cuda.synchronize()
    end_time = time.time()
    time_fp32 = end_time - start_time

    prefix = 'Group Conv2d: ' if group else 'Conv2d:'
    print(prefix + 'amp time: {:.3f}s, fp32 time: {:.3f}s, amp/fp32: {:.3f}'.format(time_amp, time_fp32, time_amp / time_fp32))


if __name__ == '__main__':
    test_with_arch(False)
    test_with_arch(True)

```

**Comment by ngimel** [comment]
That's a result on V100 GPU. @ptrblck, @mcarilli for RTX.

**Comment by buoyancy99** [comment]
> That's a result on V100 GPU. @ptrblck, @mcarilli for RTX.

Strange. I just benchmarked in V100 with Pytorch 1.6 and got 
```
Conv2d:amp time: 0.426s, fp32 time: 0.810s, amp/fp32: 0.526
Group Conv2d: amp time: 1.219s, fp32 time: 0.489s, amp/fp32: 2.490
```

I will write code with _benchmark.py and report back later

**Comment by ptrblck** [comment]
@buoyancy99 which CUDA, cudnn versions and which RTX GPU are you using?

**Comment by buoyancy99** [comment]
> @buoyancy99 which CUDA, cudnn versions and which RTX GPU are you using?

I reproduced the slow down in Tesla V100, RTX 2080TI,  Quadro RTX 6000.
There is no noticeable slow down on GTX 1080Ti, Titan Xp, in which amp/fp32 time is ~1 for both cases

CUDA version 10.2, Cudnn v7.6.5

**Comment by ngimel** [comment]
My V100 result is on cudnn 7.6.3 so perhaps there's been a regression.

**Comment by buoyancy99** [comment]
> My V100 result is on cudnn 7.6.3 so perhaps there's been a regression.
> Try setting `torch.backends.cudnn.benchmark=True`, it could improve results.

`torch.backends.cudnn.benchmark=True` helps to some extent but it still seems amp has worse performance than fp32:
```
Conv2d:amp time: 0.397s, fp32 time: 0.583s, amp/fp32: 0.681
Group Conv2d: amp time: 0.433s, fp32 time: 0.371s, amp/fp32: 1.169
```

Device: 2080Ti. Another exp on Titan RTX shows similar result

**Comment by buoyancy99** [comment]
Group conv seems to be something problematic with AMP back in apex time: found this [discussion](https://discuss.pytorch.org/t/f-conv2d-runs-x15-slower-in-mixed-precision-half-precision-mode/69987)

@ptrblck, @mcarilli anyway to fix slow down /  accelerate group convolution?
